[IR-2164] Unit Test - Physics

heysokam opened the pull request:
## Summary

## Subtasks Checklist
- [x] All `Physics.ts` Rapier->ECS APIs
- [x] ~~All `Physics.ts` External APIs~~
- [x] `RigidbodyComponent`
- [x] `ColliderComponent`
- [x] `CollisionComponent`
- [x] `TriggerComponent`
- [x] `TriggerSystem`
- [x] `PhysicsSystem`

## Breaking Changes

## References

## QA Steps

heysokam committed:
tst: UnitTests for the `Physics.ts` API (phase 0)

heysokam commented:
@HexaField @DanielBelmes
1. Are these tests going in the right direction?
2. I have tests for like 35-40% of a single file, and its already at 700sloc. Should I subdivide this task into multiple PRs? It feels massive for a single one :thinking:

HexaField commented on packages/spatial/src/physics/classes/Physics.test.ts:
```diff
@@ -361,3 +362,714 @@ describe('Physics', () => {
+
+    describe('setRigidbodyPose', () => {
+      const position = new Vector3(1, 2, 3)
+      const rotation = new Quaternion(4, 5, 6, 1)
```
we should only be using valid values, we should normalize this value

heysokam replied:
The passed quaternion was being normalized internally by Rapier.
So the issue was that I couldn't check the Expected and the Result, because they were different no matter the expected value.
Calling `.normalize()` on the input/expected quaternion, no matter their values, solved the problem :writing_hand:

HexaField commented on packages/spatial/src/physics/classes/Physics.test.ts:
```diff
@@ -361,3 +362,714 @@ describe('Physics', () => {
+    })
+
+    /**
+  // @todo How to check for the impulse of an entity?
```
impulse will have a resultant velocity the next frame, we can calculate what the velocity is, or just assert it's beforehand and assert it's not zero or close to an expected value afterwards

heysokam replied:
```ts
it("should apply the impulse to the RigidBody of the entity", () => {
  const testImpulse = new Vector3(1,2,3)
  const body = Physics._Rigidbodies.get(testEntity)
  assert.ok(body)
  
  console.log(JSON.stringify(body.linvel()))
  console.log(JSON.stringify(body.angvel()))
  
  Physics.applyImpulse(testEntity, testImpulse)
  
  console.log(JSON.stringify(body.linvel()))
  console.log(JSON.stringify(body.angvel()))
})
```
```ts
// Output
{"x":0,"y":0,"z":0}
{"x":0,"y":0,"z":0}
{"x":0,"y":0,"z":0}
{"x":0,"y":0,"z":0}
```
Am I missing something to "sync" the physics so that they are applied right away?

HexaField replied:
you need to run the physics simulation, looks at PhysicsSystem.ts

heysokam replied:
Fixed somewhere in the commit list of the PR
_Forgot to track the commit where I applied it, and now I cant find the sha_

HexaField commented on packages/spatial/src/physics/classes/Physics.test.ts:
```diff
@@ -361,3 +362,714 @@ describe('Physics', () => {
+
+    /**
+  // @todo How to check rotations?
+  describe("lockRotations", () => {
```
apply an impulse and assert it has not rotated along the locked axis

heysokam replied:
After applying the reference rotation values on creation, the resulting Quat does not remain the same compared to the starting reference. It somehow changes to float values with many decimals that have a lot of variance from run to run due to floating point imprecision.

So, this "how to check rotations" note is for asking about asserting Quaternion/Rotation results, as in how to assert the contained float values themselves in a reliable way :thinking:

heysokam replied:
Fixed somewhere in the commit list.   _(forgot to track the commit's sha)_

The mistake was that Quaterions were being normalized internally in Rapier, but I wasn't sending a normalized value.
It was solved by normalizing the input value and the checking the resulting data against that input.

HexaField commented on packages/spatial/src/physics/classes/Physics.test.ts:
```diff
@@ -361,3 +362,714 @@ describe('Physics', () => {
+
+  describe('Colliders', () => {
+    /**
+    // @todo How to check for `CollisionGroups` behavior?
```
you can use the collisionGroups() and isSensor() functions on the collider object

heysokam replied:
ty. Solved in 9e63efe3bee15f048d18d911eec260ef5f12eccb

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: UnitTests for the `Physics.ts` API (phase 1)

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: UnitTests for the `Physics.ts` API (phase 2)

heysokam committed:
fix: `RigidBodyComponent.enabledRotations` allowed type coercion

`RigidBodyComponent.onSet` allowed type modification/coercion when passing an array of `length === 3` that has incorrectly typed values.
This fix should stop `@ts-ignore` from accidentally modifying values that shouldn't be modified.

heysokam committed:
tst: UnitTests for `RigidBodyComponent.ts` (phase 0)

heysokam committed:
tst: UnitTests for `ColliderComponent.tsx` (phase 0)

heysokam committed:
tst: unitTests for `CollisionComponent.ts`

heysokam committed:
tst: Typo bugfix in `ColliderComponent.test.tsx`

heysokam committed:
tst: UnitTests for `TriggerComponent.ts` (phase 0)

heysokam committed:
tst: Files sketch for the Physics systems unit tests

heysokam committed:
tst: UnitTests for the `Physics.ts` API (phase 3)

heysokam committed:
tst: UnitTests for the `Physics.ts` API (phase 4)

heysokam committed:
tst: UnitTests for the `Physics.ts` API (phase 5)

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: UnitTests for the `Physics.ts` API (phase 6)

heysokam committed:
doc: Update `@todo` tags in `Physics.test.ts`

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: Bugfixes for `setCollisionLayer` tests at `Physics.test.ts`

heysokam committed:
tst: UnitTests for `createColliderDesc` at `Physics.test.ts` (phase 7)

heysokam committed:
tst: UnitTests for `Physics.ts` API (phase 8)

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
chg: `PhysicsState` properties: `drainCollisions` and `drainContacts`

heysokam committed:
tst: Fix UnitTests for `applyImpulse` to use new `PhysicsState` props

heysokam committed:
tst: UnitTests for `Physics.attachCollider`

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: UnitTests for `Physics.setColliderPose`

heysokam committed:
tst: UnitTests for `Physics.setEnabledRotations`

heysokam committed:
tst: UnitTests for `Physics.updatePreviousRigidbodyPose`

heysokam committed:
tst: UnitTests for `Physics.updateRigidbodyPose`

heysokam committed:
tst: Sketch for `Physics.setKinematicRigidbodyPose` UnitTests

heysokam committed:
tst: `ColliderDesc.shape.type` tests for `Physics.createColliderDesc`

heysokam committed:
tst: Sketch for `ColliderDesc.position & rotation` UnitTests

heysokam committed:
tst: UnitTests for `Physics.computeColliderMovement`

heysokam committed:
tst: UnitTests for `Physics.getComputedMovement`

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
doc: Remove leftover `@todo` tags in `Physics.test.ts`

heysokam committed:
tst: @todo list: Tests for both `Physics.drain*EventQueue` functions

heysokam committed:
tst: UnitTests for `TriggerComponent.reactor`

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: Sketch tests setup for `ColliderComponent.reactor`

heysokam committed:
tst: UnitTests for `ColliderComponent.reactor` (phase 1)

heysokam committed:
tst: UnitTests for `RigidBodyComponent.reactor` (phase 1)

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
doc: Comment on `lockRotations` and `setEnabledRotations` similarities

heysokam committed:
tst: UnitTests for `RigidBodyComponent.reactor` (phase 2)

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: UnitTests for `ColliderComponent.reactor` (phase 2)

heysokam committed:
tst: Replace hack for `getAncestorWithComponent` with `includeSelf`

heysokam committed:
tst: Small fixes on `ColliderComponent.reactor` tests

heysokam committed:
tst: UnitTests for `RigidBodyComponent.reactor` (phase 3)

heysokam committed:
git: Revert change to the return variable name of `createColliderDesc`

heysokam committed:
tst: Sketch for `TriggerSystem` UnitTests

heysokam committed:
tst: Sketch for `PhysicsSystem` UnitTests

heysokam committed:
tst: Test list: `drainCollisionEventQueue` & `drainContactEventQueue`

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
doc: @todo comments for `Physics.createColliderDesc` undefined props

heysokam committed:
tst: Bugfixes for `triggerEnter` at `TriggerSystem.test.ts`

heysokam committed:
tst: Bugfixes for multiple cases at `Physics.test.ts`

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: Bugfix for `triggerExit` UnitTests

heysokam committed:
tst: Bugfix for `Physics.drainCollisionEventQueue` UnitTests

heysokam committed:
doc: Add warning to `Physics.lockRotations` for the unlock bug

heysokam committed:
fix: Temporary fix for `Rapier < v0.12.0` bug @`Physics.lockRotations`

heysokam committed:
fix: `RigidBodyComponent.enabledRotations` not retriggering on change

heysokam committed:
tst: Export `assertFloatApprox*` functions for other Physics tests

heysokam committed:
tst: UnitTests for `RigidBodyComponent.reactor` (phase 4)

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: UnitTests for `TriggerSystem` (phase 1)

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: UnitTests for `TriggerSystem` (phase 2)

heysokam committed:
tst: UnitTests for `TriggerSystem` (phase 3)

heysokam committed:
tst: Bugfixes in `Physics.test.ts` vectors and rotations asserts

heysokam committed:
tst: UnitTests for `Physics.drainCollisionEventQueue` (phase 2)

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: UnitTests for `Physics.drainCollisionEventQueue` (phase 3)

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: UnitTests for `PhysicsSystem.smoothKinematicPose`:`multiplier 0`

heysokam committed:
tst: UnitTests for `PhysicsSystem.smoothKinematicPose`:`multiplier !0`

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
fmt: Simplify the `PhysicsState` `drain*` type declarations

heysokam committed:
tst: Add `assertVecAnyApproxNotEq` helper. Improve float assert logic

heysokam committed:
tst: Export `assertRigidBodyComponentEqual` for other tests

heysokam committed:
tst: UnitTests for `PhysicsSystem.execute`

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: UnitTests for `Physics.drainContactEventQueue`

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
fix: Hookstate CI/CD error

heysokam committed:
tst: Add @todo note about `TriggerSystem` tests for `onEnter/onExit`

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

heysokam committed:
tst: Remove outdated/redundant commented-out tests

heysokam committed:
Merge branch 'dev' into IR-2164-unit-test-physics

HexaField committed:
Merge branch 'dev' into IR-2164-unit-test-physics
